crates/astria-sequencer/src/ibc/ics20_transfer.rs

```diff
@@ -373,6 +379,42 @@ async fn execute_ics20_transfer_bridge_lock<S: StateWriteExt>(
     Ok(())
 }
 
+async fn convert_denomination_if_ibc_prefixed<S: StateReadExt>(
+    state: &mut S,
+    packet_denom: Denom,
+) -> Result<Denom> {
+    // if the asset is prefixed with `ibc`, the rest of the denomination string is the asset ID,
+    // so we need to look up the full trace from storage.
+    // see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md#decision
+    let denom = if packet_denom.prefix().starts_with("ibc") {
+        let id_bytes: [u8; 32] = hex::decode(packet_denom.base_denom())
+            .context("failed to decode ibc/ prefixed id as hex")?
+            .try_into()
+            .map_err(|_| anyhow::anyhow!("ibc/ prefixed id was not 32 bytes"))?;
+        state
+            .get_ibc_asset(id_bytes.into())
+            .await
+            .context("failed to get denom trace from asset id")?
+    } else {
+        packet_denom
+    };
+
+    Ok(denom)
+}
+
+fn prefix_denomination(
+    packet_denom: Denom,
+    dest_port: &PortId,
+    dest_channel: &ChannelId,
+    is_refund: bool,
+) -> Denom {
```

Comment:
looking at how you use this this might be a use case for `Cow<'a, Denom>` with `Cow::Borrowed(&'a Denom)` if `is_refund` and `Cow::Owned(Denom)` otherwise.